[APM] disable agent configuration creation for opentelemetry agents

achyutjhunjhunwala opened the pull request:
Closes https://github.com/elastic/kibana/issues/78802

## Summary

- The PR disables creation of APM Agent Configuration in APM Settings for services belonging to OpenTelemetry Agents.

- Regarding existing configurations, they won't be impacted if they have already been configured.

- The current service selector has an `All` option, if that is selected, the callout will not be displayed and the user will be allowed to create the Setting.

## Demo


https://user-images.githubusercontent.com/7416358/217913110-d263b311-9f6f-4a1f-9a66-b548ee02026c.mov


@boriskirov Can you have a look and provide feedback on the design.

@elastic/ux-writers Team can you help by providing proper message and title which can be displayed.

Currently used
Title - `Services with OpenTelemetry Agents are not supported`
Description - `Selected service uses an OpenTelemetry agent. OpenTelemetry agents are currently not supported by the agent configuration service.`

achyutjhunjhunwala committed:
disable agent configuration creation for opentelemetry agents

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

achyutjhunjhunwala committed:
Merge branch 'main' into restrict_agent_configuration_for_otel_services

achyutjhunjhunwala committed:
use helper function to check if agent supports open telemetry

achyutjhunjhunwala committed:
Merge branch 'main' into restrict_agent_configuration_for_otel_services

achyutjhunjhunwala committed:
chore: simplified if-else condition

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

sqren commented on x-pack/plugins/apm/public/components/app/settings/agent_configurations/agent_configuration_create_edit/service_page/service_page.tsx:
```diff
@@ -81,6 +94,14 @@ export function ServicePage({ newConfig, setNewConfig, onClickNext }: Props) {
     })
   );
 
+  useEffect(() => {
+    setIsAgentConfigurationSupported(
+      newConfig.agent_name
+        ? !isOpenTelemetryAgentName(newConfig.agent_name as AgentName)
+        : true
+    );
+  }, [newConfig.agent_name]);
```
Do this need to be stateful? I don't think it's problematic doing this in the render function:

```ts
const isAgentConfigurationSupported  = isOpenTelemetryAgentName(newConfig.agent_name as AgentName);
```
(if you really care about perf you could `useMemo` but I don't think that's needed)

achyutjhunjhunwala replied:
Ah yes, we don't need this to be stateful, i have changed the logic.
Regarding the use of `useMemo`, i don't feel it will add any added advantage as `isOpenTelemetryAgentName` is simply a `Array.includes` on an array of length close to 10, so not much of a gain.

gchaps commented:
Here is a suggestion for copy:

Title: Unable to create an agent configuration for <_name of service_>
Description: This service uses an OpenTelemetry agent, which is not supported.

Or if the service name might be long, this format might be better:

Title: Unable to create an agent configuration for the service you selected
Description: <_service name_> uses an OpenTelemetry agent, which is not supported.

boriskirov commented:
Thanks for reaching out:
1. My expectations are if this is an error because of the selection, then we display an select input error 
<img width="577" alt="image" src="https://user-images.githubusercontent.com/13353203/218062869-2d640762-110b-4546-a77b-ee39460141b4.png">
2. If we allow for "All" to be selected which would imply the OTel agents, then we can use a CallOut to communicate that all service means without the Otel instrumented one
<img width="577" alt="image" src="https://user-images.githubusercontent.com/13353203/218063148-0bcf46e1-24bc-45ca-84c5-c1b95c97bf50.png">

achyutjhunjhunwala committed:
feedback: remove unnecessary stateful setup

achyutjhunjhunwala commented:
@boriskirov @gchaps 
I have updated the demo here for your reference


https://user-images.githubusercontent.com/7416358/218078787-583d0b4b-8c8d-487b-a020-6279ff65636c.mov

achyutjhunjhunwala committed:
feedback: UX changes

achyutjhunjhunwala committed:
Merge branch 'main' into restrict_agent_configuration_for_otel_services

achyutjhunjhunwala commented:
> Question: Can should OpenTelemetry agent be singular or plural?

Singular, every service is mapped 1:1 with an agent

> Is there anything more specific we can say here? How will the configuration impact a service? That information might be a better title than "Proceed with caution."

Proceed with caution is just a placeholder I copied from the Design System at eui.elastic.co
Regarding the impact, when `All` is selected, only Non Telemetry supported agents will have effect with this configuration. Otel agents will have no change at all.

achyutjhunjhunwala committed:
updated error/warning messages

achyutjhunjhunwala committed:
Merge branch 'main' into restrict_agent_configuration_for_otel_services

achyutjhunjhunwala commented:
After discussion with @gchaps, i have removed the Callout Title and replaced the message as discussed on Slack.

## Error Situation

<img width="1237" alt="image" src="https://user-images.githubusercontent.com/7416358/220165748-192e4df7-1e7f-4821-b90c-62f5ea44aac5.png">


## All Option Selected Situation

<img width="1253" alt="image" src="https://user-images.githubusercontent.com/7416358/220165837-ef3d7f3c-0648-4c60-b383-e00b0e356053.png">

yngrdyn commented on x-pack/plugins/apm/public/components/app/settings/agent_configurations/agent_configuration_create_edit/service_page/form_row_suggestions_select.tsx:
```diff
@@ -23,6 +23,8 @@ interface Props {
   onChange: (value?: string) => void;
   dataTestSubj?: string;
+  isInvalid?: boolean;
+  error?: ReactNode | ReactNode[];
```
Why do we have a `ReactNode[]` here? Shouldn't be a single error?

achyutjhunjhunwala replied:
The EUI component `EuiComboBox` is the ultimate receiver of this props and it expects both option. Since this is a generic component, i made it expect the same.

achyutjhunjhunwala committed:
Merge branch 'main' into restrict_agent_configuration_for_otel_services

yngrdyn commented on x-pack/plugins/apm/public/components/app/settings/agent_configurations/agent_configuration_create_edit/service_page/service_page.tsx:
```diff
@@ -162,7 +201,8 @@ export function ServicePage({ newConfig, setNewConfig, onClickNext }: Props) {
               !newConfig.service.environment ||
-              agentNameStatus === FETCH_STATUS.LOADING
+              agentNameStatus === FETCH_STATUS.LOADING ||
+              !isAgentConfigurationSupported
```
[nit] this is growing, could we maybe move it to a const?

achyutjhunjhunwala replied:
Fixed

achyutjhunjhunwala committed:
Merge branch 'main' into restrict_agent_configuration_for_otel_services

achyutjhunjhunwala committed:
update nit comment and warning message

gchaps reviewed (approved):
text LGTM
